ocaml/typing/solver.ml

```diff
@@ -0,0 +1,779 @@
+      | Amode c -> Amode c
+      | Amodevar mv -> Amodevar (allow_right_mv mv)
+      | Amodemeet (c, mvs) -> Amodemeet (c, List.map allow_right_mv mvs)
+
+    and allow_right_mv :
+        type a l r. (a, l * allowed) morphvar -> (a, l * r) morphvar = function
+      | Amorphvar (v, m) -> Amorphvar (v, C.allow_right m)
+
+    let rec disallow_left :
+        type a l r. (a, l * r) mode -> (a, disallowed * r) mode = function
+      | Amode c -> Amode c
+      | Amodevar mv -> Amodevar (disallow_left_mv mv)
+      | Amodejoin (c, mvs) -> Amodejoin (c, List.map disallow_left_mv mvs)
+      | Amodemeet (c, mvs) -> Amodemeet (c, List.map disallow_left_mv mvs)
+
+    and disallow_left_mv :
+        type a l r. (a, l * r) morphvar -> (a, disallowed * r) morphvar =
+      function
+      | Amorphvar (v, m) -> Amorphvar (v, C.disallow_left m)
+
+    let rec disallow_right :
+        type a l r. (a, l * r) mode -> (a, l * disallowed) mode = function
+      | Amode c -> Amode c
+      | Amodevar mv -> Amodevar (disallow_right_mv mv)
+      | Amodejoin (c, mvs) -> Amodejoin (c, List.map disallow_right_mv mvs)
+      | Amodemeet (c, mvs) -> Amodemeet (c, List.map disallow_right_mv mvs)
+
+    and disallow_right_mv :
+        type a l r. (a, l * r) morphvar -> (a, l * disallowed) morphvar =
+      function
+      | Amorphvar (v, m) -> Amorphvar (v, C.disallow_right m)
+  end)
+
+  let mlower dst (Amorphvar (var, morph)) = C.apply dst morph var.lower
+
+  let mupper dst (Amorphvar (var, morph)) = C.apply dst morph var.upper
+
+  let min (type a) (obj : a C.obj) = Amode (C.min obj)
+
+  let max (type a) (obj : a C.obj) = Amode (C.max obj)
+
+  let of_const a = Amode a
+
+  let apply_morphvar dst morph (Amorphvar (var, morph')) =
+    Amorphvar (var, C.compose dst morph morph')
+
+  let apply :
+      type a b l r.
+      b C.obj -> (a, b, l * r) C.morph -> (a, l * r) mode -> (b, l * r) mode =
+   fun dst morph -> function
+    | Amode a -> Amode (C.apply dst morph a)
+    | Amodevar mv -> Amodevar (apply_morphvar dst morph mv)
+    | Amodejoin (a, vs) ->
+      Amodejoin (C.apply dst morph a, List.map (apply_morphvar dst morph) vs)
+    | Amodemeet (a, vs) ->
+      Amodemeet (C.apply dst morph a, List.map (apply_morphvar dst morph) vs)
+
+  let update_lower (type a) ?log (obj : a C.obj) v a =
+    (match log with
+    | None -> ()
+    | Some log -> log := Clower (v, v.lower) :: !log);
+    v.lower <- C.join obj v.lower a;
+    if not (C.le obj v.lower v.upper)
+    then (
+      Format.eprintf "range insane after update_lower: %a\n" (print_var obj) v;
+      assert false)
+
+  let update_upper (type a) ?log (obj : a C.obj) v a =
+    (match log with
+    | None -> ()
+    | Some log -> log := Cupper (v, v.upper) :: !log);
+    v.upper <- C.meet obj v.upper a;
+    if not (C.le obj v.lower v.upper)
+    then (
+      Format.eprintf "range insane after update_lower: %a\n" (print_var obj) v;
+      assert false)
+
+  let set_vlower ?log v vlower =
```

Comment:
`lower` is a conservative bound and is not updated even if the `lower` of some variable in `vlower` is updated. 

For example, if `v.vlower` is a single `u`, and `u.lower` is `5`; therefore, `v.lower` should also be `5`. Now if we update `u.lower` to be `6`, this will not propagate to `v.lower` immediately.